java/arcs/core/analysis/ClaimDeduction.kt

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+package arcs.core.analysis
+
+import arcs.core.data.expression.Expression
+
+internal typealias Path = List<String>
+internal typealias Paths = List<Path>
+internal typealias ClaimDerivations = Map<Path, Set<Path>>
+
+fun <E> MutableList<E>.push(element: E) = add(element)
+fun <E> MutableList<E>.pop(): E = removeAt(size - 1)
+fun <E> MutableList<E>.peek(): E = this[count()]
+fun <E> MutableList<E>.popOrDefault(default: E): E = if (size == 0) default else pop()
+fun <E> MutableList<E>.peekOrDefault(default: E): E = if (size == 0) default else peek()
+
+/** A visitor that accumulates [ClaimDerivations] from a Paxel [Expression]. */
+class ClaimDeducer : Expression.Visitor<ClaimDerivations> {
```

Comment:
I believe you want the Visitor to return 2 things (tuple/data-class with 2 fields):

1) Derivations of the returned value (if any).
2) Map from identifiers introduced into the scope to their derivations.

I think all our expressions either introduce a new variable into the scope (`from...in`, `let x = ...`), or return a value and close the scope (`select ...`, any binary / unary operator), so we could in theory return one or the other if we had a nice way to represent that in Kotlin.

For example:
`<expr1> + <expr2>` will return a value that is derived from values of expr1 and expr2, and will have an empty map of values it introduces to the scope.

`from f in foo.x` does not return any value by itself, so the derivation of the returned value is null, but it returns a map of `{f: foo.x}`, signifying that the identifier `f` derives from `foo.x`.

A more complex example for introducing things into the scope:

```
from f in foo
from b in bar
let x = f.x + b.y
let y = x * first(baz).z
```

will return `null` as the derivation of the value it returns (expression does not return anything), but returns a map of:

```
{
  f: [foo],
  b: [bar],
  x: [foo.x, bar.y]
  y: [foo.x, bar.y, baz.z]
}
```

@bgogul to comment, he may have some ideas.

Reply:
Ray will be refactoring all the expressions that modify the currentScope to return Scope or Sequence<Scope>, which simplifies things a lot, as you only need a visitor that returns derivations claims for the returned value.

Reply:
Piotr and I had a conversation about this. The takeaway: Given some refactoring that Ray is working on for OrderBy, we can likely get away with the strategy in this PR (i.e. we won't have to return a tuple).